Sources/FaceLiveness/Views/Liveness/FaceLivenessDetectionView.swift

```diff
@@ -226,12 +224,32 @@ public struct FaceLivenessDetectorView: View {
             break
         }
     }
-}
-
-enum CountdownDisplayState {
-    case waitingToDisplay
-    case displaying
-    case finishedDisplaying
+    
+    func mapErrorToCloseCode(_ livenessError: LivenessStateMachine.LivenessError) -> URLSessionWebSocketTask.CloseCode {
+        var closeCode: URLSessionWebSocketTask.CloseCode?
+        switch livenessError {
+        case .userCancelled:
+            closeCode = URLSessionWebSocketTask.CloseCode.ovalFitUserClosedSession
+        case .timedOut:
+            closeCode = URLSessionWebSocketTask.CloseCode.ovalFitMatchTimeout
+        case .socketClosed:
+            closeCode = .normalClosure
+        case .missingVideoPermission:
+            closeCode = URLSessionWebSocketTask.CloseCode.missingVideoPermission
+        case .viewResignation:
+            closeCode = URLSessionWebSocketTask.CloseCode.viewClosure
+        case .unknown, .errorWithUnderlyingOSFramework, .couldNotOpenStream:
+            closeCode = URLSessionWebSocketTask.CloseCode.unexpectedRuntimeError
+        default:
+            closeCode = .normalClosure
+        }
+        
+        if let code = closeCode {
+            return code
+        } else {
+            return URLSessionWebSocketTask.CloseCode.normalClosure
+        }
+    }
```

Comment:
Looks good. One (non-blocking) suggestion - consider adding the websocket close frame code to the `LivenessStateMachine.LivenessError` struct rather than mapping in the switch statement.

```swift
struct LivenessError: Error, Equatable {
    let code: UInt8
    let webSocketCloseCode: URLSessionWebSocketTask.CloseCode?

    static let unknown = LivenessError(
        code: 0,
        webSocketCloseCode: .unexpectedRuntimeError
    )

    static let missingVideoPermission = LivenessError(
        code: 1,
        webSocketCloseCode: .missingVideoPermission
    )

    static let errorWithUnderlyingOSFramework = LivenessError(
        code: 2,
        webSocketCloseCode: .unexpectedRuntimeError
    )

    static let userCancelled = LivenessError(
        code: 3,
        webSocketCloseCode: .ovalFitUserClosedSession
    )

   // ...

    static func == (lhs: LivenessError, rhs: LivenessError) -> Bool {
        lhs.code == rhs.code
    }
}
```

Which will let you:
```swift
viewModel.livenessService.closeSocket(with: error.webSocketCloseCode ?? .normalClosure)
```

Reply:
I've updated the code per suggestion, it does look a lot cleaner!